cit3.0-web/src/hooks/useKeycloakWrapper.js

```diff
@@ -0,0 +1,185 @@
+//   lastName?: string;
+//   email?: string;
+//   roles: string[];
+//   agencyId?: number;
+//   isAdmin: boolean;
+//   hasRole(role?: string | Array<string>): boolean;
+//   hasClaim(claim?: string | Array<string>): boolean;
+//   hasAgency(agency?: number): boolean;
+//   agencyIds: number[];
+//   canUserEditProperty: (property: IProperty | null) => boolean;
+//   canUserViewProperty: (property: IProperty | null) => boolean;
+//   canUserDeleteProperty: (property: IProperty | null) => boolean;
+// }
+
+/**
+ * Provides extension methods to interact with the `keycloak` object.
+ */
+export function useKeycloakWrapper() {
+  const { keycloak } = useKeycloak();
+  const userInfo = keycloak && keycloak.userInfo;
+
+  /**
+   * Determine if the user has the specified 'claim'
+   * @param claim - The name of the claim
+   */
+  const hasClaim = (claim) =>
+    claim !== undefined &&
+    claim !== null &&
+    userInfo &&
+    userInfo.roles &&
+    (typeof claim === "string"
+      ? userInfo.roles.includes(claim)
+      : claim.some((c) => userInfo.roles.includes(c)));
+
+  /**
+   * Determine if the user belongs to the specified 'role'
+   * @param role - The role name or an array of role name
+   */
+  const hasRole = (role) =>
+    role !== undefined &&
+    role !== null &&
+    userInfo &&
+    userInfo.groups &&
+    (typeof role === "string"
+      ? userInfo.groups.includes(role)
+      : role.some((r) => userInfo.groups.includes(r)));
+
+  /**
+   * Determine if the user belongs to the specified 'agency'
+   * @param agency - The agency name
+   */
+  const hasAgency = (agency) =>
+    agency !== undefined &&
+    agency !== null &&
+    userInfo &&
+    userInfo.agencies &&
+    userInfo.agencies.includes(agency);
+
+  /**
+   * Return an array of roles the user belongs to
+   */
+  const roles = () => (userInfo && userInfo.groups ? [...userInfo.groups] : []);
+
+  /**
+   * Return the user's username
+   */
+  const username = () => userInfo && userInfo.username;
+
+  /**
+   * Return the user's display name
+   */
+  const displayName = () =>
+    userInfo && (userInfo.name || userInfo.preferred_username);
+
+  /**
+   * Return the user's first name
+   */
+  const firstName = () =>
+    userInfo && (userInfo.firstName || userInfo.given_name);
+
+  /**
+   * Return the user's last name
+   */
+  const lastName = () =>
+    userInfo && (userInfo.lastName || userInfo.family_name);
+
+  /**
+   * Return the user's email
+   */
+  const email = () => userInfo && userInfo.email;
+
+  /**
+   * Return true if the user has permissions to edit this opportunity
+   */
+  const canUserEditOpportunity = (property) =>
+    !!property &&
+    (hasClaim(Claims.ADMIN_PROPERTIES) ||
+      (hasClaim(Claims.PROPERTY_EDIT) &&
+        !!property.agencyId &&
+        hasAgency(property.agencyId) &&
+        !_.some(
+          property.projectNumbers ? property.projectNumbers : "",
+          _.method("includes", "SPP")
+        )));
```

Comment:
This could maybe use an explanatory comment for the conditions that allow a user to edit the opportunity, or be split up a bit to be more readable. This is a pretty lengthy boolean expression.

Reply:
Same comment for the two functions below that are similar.

Reply:
This is not final form, its an example from PIMS. As we don't ahve the oidc tentant provisioned, i decided not to even bother changing this yet. I will just add a todo comment.

Reply:
Sounds good!